gapis/resolve/dependencygraph/footprint.go

```diff
@@ -0,0 +1,239 @@
+type BackPropagationMachine interface {
+	// IsAlive check whether a given behaviour, which is specified by its index
+	// and the footprint to which it belongs, should be kept alive. Returns true
+	// if the behaviour should be kept alive, otherwise returns false.
+	IsAlive(behaviourIndex uint64, ft *Footprint) bool
+	// RecordBehaviourEffects records the read/write operations of the given
+	// behaviour specified by its index in the given footprint, and returns the
+	// indices of the alive behaviours that should be kept alive due to recording
+	// of the given behaviour.
+	RecordBehaviourEffects(behaviourIndex uint64, ft *Footprint) []uint64
+	// Clear clears the internal state of the BackPropagationMachine.
+	Clear()
+}
+
+// dummyMachine does nothing but marks all the incoming Behaviours as alive.
+type dummyMachine struct{}
+
+func (m *dummyMachine) IsAlive(uint64, *Footprint) bool { return true }
+func (m *dummyMachine) RecordBehaviourEffects(behaviourIndex uint64,
+	ft *Footprint) []uint64 {
+	return []uint64{behaviourIndex}
+}
+
+func (m *dummyMachine) Clear() {}
+
+// GetBehaviourIndex returns the index of the last Behaviour in the Footprint
+// which belongs to the command or subcomand indexed by the given SubCmdIdx. In
+// case the SubCmdIdx is invalid or the Behaviour index is not found, error
+// will be logged and uint64(0) will be returned.
+func (fpt *Footprint) GetBehaviourIndex(ctx context.Context,
+	fci api.SubCmdIdx) uint64 {
+	switch len(fci) {
+	case 1,
+		// Only API command ID
+		4,
+		// For Vulkan: ID of VkQueueSubmit -> SubmitInfo Index ->
+		// CommandBuffer Index -> Command Index
+		6:
+		// For Vulkan: ID of VkQueueSubmit -> SubmitInfo Index ->
+		// CommandBuffer Index -> Index of vkCmdExecuteCommands ->
+		// secondary CommandBuffer Index -> Command Index
+		v := fpt.commandIndexToBehaviourIndexLookupTable.Value(fci)
+		if u, ok := v.(uint64); ok {
+			return u
+		} else {
+			log.E(ctx, "Invalid value of behaviour index: %v, (not a uint64)", v)
+			return uint64(0)
+		}
+	default:
+		log.E(ctx, "Invalid length of SubCmdIdx (full command index): %v", len(fci))
+		return 0
+	}
+}
+
+// AddBehaviour adds the given Behaviour to the Footprint and updates the
+// internal mapping from SubCmdIdx to the last Behaviour that belongs to that
+// command or subcommand.
+func (fpt *Footprint) AddBehaviour(ctx context.Context, b *Behaviour) bool {
+	bi := uint64(len(fpt.Behaviours))
+	fci := b.BelongTo
+	switch len(fci) {
+	case 1, 4, 6:
+		fpt.commandIndexToBehaviourIndexLookupTable.SetValue(fci, bi)
+		fpt.Behaviours = append(fpt.Behaviours, b)
+		fpt.BehaviourIndices[b] = bi
+		return true
+	default:
+		log.E(ctx, "Invalid length of SubCmdIdx (full command index): %v", len(fci))
+		return false
+	}
+}
+
+// FootprintBuilderAPI provides FootprintBuilder
+type FootprintBuilderAPI interface {
+	FootprintBuilder(context.Context) FootprintBuilder
+}
+
+// FootprintBuilder builds Footprint from a stream of commands.
```

Comment:
`from a stream of commands` I only see one command being passed to the function. Please clarify.